ktor-server/ktor-server-plugins/ktor-server-auth/jvmAndNix/src/io/ktor/server/auth/AuthenticationInterceptors.kt

```diff
@@ -63,109 +49,178 @@ private val logger = KtorSimpleLogger("Authentication")
-                }
+private suspend fun executeChallenges(context: AuthenticationContext, call: ApplicationCall) {
+    val challenges = context.challenge.challenges
+
+    for (challenge in challenges) {
+        challenge(context.challenge, call)
+        if (context.challenge.completed) {
+            if (!call.isHandled) {
+                call.respond(UnauthorizedResponse())
             }
+            return
+        }
+    }
 
-            for (challenge in context.challenge.errorChallenges) {
-                challenge(context.challenge, call)
-                if (context.challenge.completed) {
-                    if (!call.isHandled) {
-                        call.respond(UnauthorizedResponse())
-                    }
-                    return@on
-                }
+    for (challenge in context.challenge.errorChallenges) {
+        challenge(context.challenge, call)
+        if (context.challenge.completed) {
+            if (!call.isHandled) {
+                call.respond(UnauthorizedResponse())
             }
+            return
+        }
+    }
 
-            for (error in context.allErrors) {
-                if (!context.challenge.completed) {
-                    logger.trace("Responding unauthorized because of error ${error.message}")
-                    if (!call.isHandled) {
-                        call.respond(UnauthorizedResponse())
-                    }
-                    context.challenge.complete()
-                    return@on
-                }
+    for (error in context.allErrors) {
+        if (!context.challenge.completed) {
+            logger.trace("Responding unauthorized because of error ${error.message}")
+            if (!call.isHandled) {
+                call.respond(UnauthorizedResponse())
             }
+            context.challenge.complete()
+            return
         }
     }
+}
+
+private fun AuthenticationConfig.findProviders(
+    configurations: Collection<AuthenticateProvidersRegistration>,
+    filter: (AuthenticationStrategy) -> Boolean
+): Set<AuthenticationProvider> {
+    return configurations.filter { filter(it.strategy) }
+        .flatMap { it.names.map { configurationName -> this.findProvider(configurationName) } }
+        .toSet()
+}
+
+private fun AuthenticationConfig.findProvider(configurationName: String?): AuthenticationProvider {
+    return providers[configurationName] ?: throw IllegalArgumentException(
+        if (configurationName == null) {
+            "Default authentication configuration was not found"
+        } else {
+            "Authentication configuration with the name $configurationName was not found"
+        }
+    )
+}
+
+/**
+ *  A resolution strategy for nested authentication providers.
+ *  [AuthenticationStrategy.Optional] - if no authentication is provided by the client,
+ *  a call continues but with a null [Principal].
+ *  [AuthenticationStrategy.FirstSuccessful] - client must provide authentication for at least one provider
```

Comment:
`must provide authentication` -> `must provide authentication credentials` or `must provide authentication data`
Please update here and in other places as well